src/libmain/progress-bar.cc

```diff
@@ -365,6 +368,62 @@ class ProgressBar : public Logger
         writeToStderr("\r" + filterANSIEscapes(line, false, width) + ANSI_NORMAL + "\e[K");
     }
 
+		void draw_multi(State & state)
+		{
+			state.haveUpdate = false;
+			if (!state.active) return;
+
+			std::string line;
+			auto width = getWindowSize().second;
+			if (width <= 0) width = std::numeric_limits<decltype(width)>::max();
+
+			std::string status = getStatus(state);
+			if (!status.empty()) {
+				line += '[';
+				line += status;
+				line += "]";
+			}
+
+      #if defined(WIN32) || defined(_WIN32) || defined(__WIN32__) || defined(__NT__)
+			  system("cls");
+			#else
+        writeToStderr("\r\033[K\r");
+	      for (auto i = 1; i < state.lastLines; i++)
+	        writeToStderr("\033[1A\r\033[K\r");
```

Comment:
```suggestion
        writeToStderr(fmt("\033[%dA\r\033[K", state.lastLines - 1));
```

What about using [DECSTBM](https://www.vt100.net/docs/vt510-rm/DECSTBM.html) instead? This would also get rid of line tracking since all you need to do is set the scrolling region and push new lines.

For example like this (given that `lineStart` and `lineEnd` point to the absolute top and bottom margins):

```suggestion
	      writeToStderr(fmt("\033[s\033[%d;%dr\033[u", lineStart, lineEnd));
```

The `\e[s` and `\e[u` CSI are to save and restore cursor positions.

[Here is a small demo](https://gist.github.com/aszlig/f48026a6ec1f58efad84b41e336521d7) on how this could be even improved further, which also uses [DECSLRM](https://www.vt100.net/docs/vt510-rm/DECSLRM.html) if the terminal supports it (currently it's just checks if `TERM` starts with `xterm`) to set left margins. The latter could also be useful for constraining build outputs (eg. in conjunction with [DECOM](https://vt100.net/docs/vt510-rm/DECOM)) to the output of prefixed status like `building somedrv: some build output` and should make a lot of the CSI handling in ProgressBar obsolete.

Running this looks something like this. Note that I deliberately filled the terminal with stuff, so that behaviour is clear when we're reaching the bottom row of the terminal:

![decstbm](https://user-images.githubusercontent.com/192147/96205189-442fdb80-0f66-11eb-8f91-3b56ff131353.gif)

The demo currently doesn't handle `SIGWINCH` and apart for the mentioned XTerm check it doesn't query terminal features. This is btw. something which I think we should probably do in Nix in general, since for example right now we just assume that SGR is *always* working on all terminals.

Reply:
@aszlig  I don't know enough about that to implement it, so feel free to take this pr over. I also would be available on https://jitsi.nixcon.net/progress-bar at the moment